### ISSUE TITLE
Add support for more AWS events

### ISSUE THREAD

**Issue by eahefnawy** [serverless/serverless#1409]
- [x] SNS (#1476 / #1499)
- [x] Streams (e.g. Kinesis and DynamoDB) (#1608 / #1441)
- [ ] SES (#3401)
- [x] Cognito (#3400)
- [x] CloudWatch logs (#3399)
- [x] CloudWatch events (#3102)
- [x] Alexa Skills Kit (#2875 / #1525)
- [x] Alexa Smart Home (#3955)
- [x] IoT (#2954)

Overview of available events for lambda:
http://docs.aws.amazon.com/lambda/latest/dg/invoking-lambda-function.html

**Comment by rajington** [comment]
"Echo" is actually "Alexa Skills Kit" (simple policy) and "Alexa Smart Home" (more complicated/hidden). I need the former right now so I can submit a PR (against 1.0) unless someone else is working on it.

**Comment by pmuens** [comment]
@rajington We're currently not working on this. Thanks for jumping in on this and submitting a PR!

**Comment by pouledodue** [comment]
DynamoDB Streams would be a good start, when one want to implement EventSourcing/CQRS

**Comment by andymac4182** [comment]
I am looking at CloudWatch Logs firing lambda for some of our internal applications. I will get the work done for a PR.

**Comment by ezjacob-maheim** [comment]
@andymac4182 I'm looking for this event as well. Is there any update on when it would added?

**Comment by sidewinder12s** [comment]
Anyone aware of any guides that might show how I could use CF to attach CloudWatch Event Rules to functions? I am looking to kick off Lambda's based on API actions like Instance Launch, SnapshotCreation, etc. Right now I have a specific router function that is listening to an S3 bucket that is getting cloudtrail delivered to it.

**Comment by rowanu** [comment]
@sidewinder12s I'm planning to do this in a week or so. Will send through a PR then if it hasn't been addressed.

**Comment by andymac4182** [comment]
@ezjacob-maheim Sorry for the late reply I have done some initial work in a plugin that I am discussing in #2621

**Comment by flomotlik** [comment]
Going to close this issue as we're not really using it at the moment for discussions. We'll work on new events when they are coming up through feedback of the community.

**Comment by nikgraf** [comment]
I propose we either either create separate issues for this or integrate SAM.

**Comment by horike37** [comment]
I will implement IOT events and send PR. Using CF.

**Comment by pmuens** [comment]
@horike37 awesome! 💯  🎉 

Looking forward to it!

**Comment by mazzy89** [comment]
I could take CloudWatch logs event because we met the need in our project to use it. As workaround we used a CF template. I'll send a PR.

**Comment by pmuens** [comment]
@mazzy89 awesome! Looking forward to the PR!

**Comment by nikgraf** [comment]
@mazzy89 that would be great! looking forward to the PR 🙌

**Comment by horike37** [comment]
@pmuens @nikgraf 
Just implemented CloudWatch event. Please review it.

**Comment by pmuens** [comment]
Woohoo! @horike37 you rock!

Will review it today!

**Comment by horike37** [comment]
Hey @mazzy89!
How is working for CloudWatch logs event?

**Comment by mazzy89** [comment]
@horike37 It will be prepared in few days ;-)

**Comment by mina-qlik** [comment]
@mazzy89 was CloudWatch logs event added? If yes, can you please forward a sample of serverless.yml on how to use it?

**Comment by sidewinder12s** [comment]
@mina-qlik https://serverless.com/framework/docs/providers/aws/events/cloudwatch-event/

**Comment by horike37** [comment]
@mina-qlik 
 CloudWatch logs event has not supported yet.

**Comment by love8587** [comment]
@mazzy89 I'm wating for CloudWatch logs! please!  :)

**Comment by horike37** [comment]
@love8587 
We have been working for CloudWatch logs on the following PR now.
https://github.com/serverless/serverless/pull/3407

We'll release it on v1.11. Just a second.:bowtie:

**Comment by love8587** [comment]
@horike37 many thanks! I've confirmed it and worked well. Good job!

**Comment by horike37** [comment]
Thank you for this comments @love8587 !
We are glad to hear it :bowtie:

**Comment by PMudra** [comment]
I have added a feature request to the [awslabs/serverless-application-model](https://github.com/awslabs/serverless-application-model) requesting support for the Alexa Smart Home Skill to be added to the list of event source types.

awslabs/serverless-application-model#158

I believe this will enable the serverless framework to add support for the Alexa Smart Home Skill as event type.

**Comment by pmuens** [comment]
Thanks for opening that issue @PMudra and cross-linking it here 👍 

Would be nice if you could keep us posted here once AWS decided to add this functionality. Best would be if it's supported through raw CloudFormation since Serverless doesn't support SAM templates yet.

**Comment by alexcasalboni** [comment]
Any plan to support also Kinesis Data Firehose and CodeCommit?

**Comment by horike37** [comment]
>Any plan to support also Kinesis Data Firehose and CodeCommit?

Of course :+1:  We can accept PRs for any events feature without prior discussions.
It can be done if those events is supported by CloudFormation.

**Comment by jvarho** [comment]
@alexcasalboni CodeCommit is already usable as an event through CloudWatch Events. E.g.:

    hello:
        handler: handler.hello
        events:
            - cloudwatchEvent:
                event:
                    source:
                        - "aws.codecommit"
                    detail-type:
                        - "CodeCommit Repository State Change"
                    resources:
                        - "arn:aws:codecommit:..." # Repository ARN here

Would there be an advantage in supporting it directly?

**Comment by alexcasalboni** [comment]
@jvarho thanks for the heads up!

Well, then we could make it much easier to find [here](https://serverless.com/framework/docs/providers/aws/) and [here](https://serverless.com/framework/docs/providers/aws/events/cloudwatch-event/) :)

**Comment by marcelobern** [comment]
@alexcasalboni @horike37 I have been working on code to support CodeCommit natively, as I ran into some issues using the CloudWatch Events approach pointed out by @jvarho 

The PR would include:
- Event processing - 100% code coverage
- Event documentation - behavior inspired on SNS event behavior
- Templates (git client & CodeCommit API) - include minimal working lambda code to help with the learning curve and putting things together

I have it working as a local plugin and will start porting into `serverless`, so please let me know in case this is not relevant anymore, in which case I will just keep it as a local plugin (or maybe a public one).

**Comment by jvarho** [comment]
> I have been working on code to support CodeCommit natively, as I ran into some issues using the CloudWatch Events approach pointed out by @jvarho 

Out of curiosity, what issues were they? I can remember being annoyed with the events when I last worked with them but cannot remember what was wrong with them.

What did you improve in the way your PR allows working with CodeCommit? I tried reading the changes, but at least the documentation did not seem to show anything that was not possible before.

**Comment by marcelobern** [comment]
@jvarho I do not remember exactly what was not working for me as it has been a few weeks/months since I tried the CloudWatch event for CodeCommit approach.

So, I just looked over the [CloudWatch events for CodeCommit](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/EventTypes.html#codecommit_event_type) to see if it would jog my memory allowing me to give you and answer :-)

After that, I seem to remember it was related to branch names and when the Lambda function was triggered. While the CloudWatch event approach does notify you of "when" it happens, it does not allow you to filter for just a few branches, thus resulting in potentially increased number of times your lambda function is invoked (unnecessarily).

I might be completely off but that is my best guess at this point.

Hopefully this helps you!
